### ISSUE TITLE
Average slope calculation

### ISSUE THREAD

**Issue by MichaelPil** [vcloarec/ReosProject#18]
Greetings,

Could you please explain how the average slope of the stream (used for the calculation of concentration time) is calculated ?

We have been experimenting with other tools, such as QGIS and its plugins, and also with some average slope calculation formulas on excel spreadsheets. The results we get on the average slope calculations using these methods, significantly differ from the slope calculated on the same stream by Lekan. Lekan always calculates almost 10% (in absolute value, for example 3.16% vs 14.28% on QGIS or by "hand") lower average slope values, meaning also higher concentration times and lower flow rates. Note that we use the same flow path on the same DEM on both Lekan and QGIS.

In fact we also copied the distance and altitude values on the Longitudinal Profile window to our spreadsheet and calculated the average slope, which didn't match to the average slope by Lekan. 

So our goal is to evaluate the accuracy of our spreadsheets and of the QGIS tools we use. Or perhaps there is some kind of error on the formula used by Lekan?

At the moment we are using Lekan version 2.2.1.

Thank you very much
Kind regards,

**Comment by vcloarec** [comment]
Hi,
Can you put here a screen shot of the longitudinal profile in Lekan with the table?

**Comment by vcloarec** [comment]
The formula used in Lekan is this one:
![averageSlope_2](https://user-images.githubusercontent.com/7416892/163576979-4b675e4f-0080-4f38-a506-db57000d72b5.svg)

where 
Li is the length of each segment
si is slope of each segment
s is the average slope

**Comment by MichaelPil** [comment]
This is the Longitudinal profile (it usually prints the slope % of its segment but for some reason this time it doesn't):
![image](https://user-images.githubusercontent.com/103752041/163576564-21e49568-4bea-4e46-a626-59a1b1f996de.png)

And here is a snapshot of the table :
![image](https://user-images.githubusercontent.com/103752041/163576945-1df8f8f0-1d03-4f39-9208-a5cdaf3386fc.png)

This is one of the watersheds where we observe the differences. In this one for example Lekan calculates 2.40 % average slope, while with other methods I calculated 12 - 14 % average slope.

**Comment by vcloarec** [comment]
Some strange stuff...
Normally, the direction of the stream is from left to right in the plot.
did you draw the stream by hand ?

**Comment by MichaelPil** [comment]
Yes I did draw it by hand because the DEM was not very precise and the software couldn't determine the flow path automatically.

And perhaps I drew it from downstream to usptream.

**Comment by vcloarec** [comment]
> Yes I did draw it by hand because the DEM was not very precise and the software couldn't determine the flow path automatically.

better to do in the other side to have a better behavior of Lekan. Anyway, this should not explain the difference.

Be sure you approach in the same as the one with this formula

> The formula used in Lekan is this one:
> 
> ![averageSlope_2](https://user-images.githubusercontent.com/7416892/163577218-afa7e00f-45c7-4dce-84dd-b35095c7362f.svg)

**Comment by MichaelPil** [comment]
This is another stream, also drawn by hand, but from upstream to downstream.
![image](https://user-images.githubusercontent.com/103752041/163578233-fce2ceb9-c531-4042-86f7-e323dd20ee65.png)

And its table:
![image](https://user-images.githubusercontent.com/103752041/163578459-cdaba3a5-837e-4e65-8e1e-45213e974159.png)

The average slope calculated is 10.68% and with other methods we calculate 20%. But I 'm not sure which formula qgis algorithms use.

**Comment by vcloarec** [comment]
which tool do you use with QGIS?

**Comment by MichaelPil** [comment]
I use the qprof plugin with the "drape" and "extract Z values" commands.

**Comment by MichaelPil** [comment]
For our spreadsheets we use 2 simple formulas :

1. s average = si*wi,
where wi is the percentage of each line segment on the full path

2. s = (Hmax-Hmin) / ΣLi

Could you please give me a citation (reference) for the formula Lekan uses?

**Comment by vcloarec** [comment]
It is a commonly used formula used for average slope, it can be named formula of Taylor Schwartz.

**Comment by MichaelPil** [comment]
Ok , Thank you very much for the information.

May I suggest an addition for newer versions? Perhaps it would be useful to add a few more formulas for the average slope calculation and give the user the option to select the preferred formula. It seems to me that Taylor Schwartz model is calculating a little milder slopes than I would expect, so it would be useful to be able to compare it with other models and choose one in a way similar to choosing the appropriate concentration time formulas.

Once again thank you for your response and congratulations for your great work and software!

Kind regards

**Comment by vcloarec** [comment]
Yes, it could be a improvement, but not the priority for now. If you need it, don't hesitate to sponsor it to make it a priority :smiley:

Note that it is not mandatory to use the slope calculated by Lekan, you can override this slope by your own.

Thanks for your kinds words and don"t hesitate to help if you like Lekan: https://www.reos.site/en/how-to-support/

**Comment by pertetotale** [comment]
sorry to be a bit late: 
Did you fill sinks on DEM? 
Also I noticed your pic with enlarged slope distance.

But the expr. I got with DEMs of "less interesting locations":  there might be area's that are not scanned 3*, only 2*, 
so we might miss accurate slopes in certain angles. 

Btw, Vincent, I remember an error in a formula which showed in a pop up: exp x -y vs. exp +y, which seemed to me caused by an [latex] ()/{} misplacement, so I hoped it would be correctly implemented and I did not signal it. But I did not look closer in it. 

On a good note, I did a Green Ampt run, and it seemed to be even a bit better, but you really need to have decent soil info, (depth, infil, soil mixes, the layers...) and general soil type insight.

Just for your info.
